library/core/src/pin.rs

```diff
@@ -1246,5 +1237,5 @@ pub macro pin($value:expr $(,)?) {
     //
     // See https://doc.rust-lang.org/1.58.1/reference/destructors.html#temporary-lifetime-extension
     // for more info.
-    $crate::pin::Pin::<&mut _> { pointer: &mut { $value } }
+    Pin { pointer: &mut { $value } }
```

Comment:
Removal of turbofish type is orthogonal to field privacy / hygiene, and technically allows for code such as `let p: Pin<*mut _> = pin!(...);` (letting a coërcion happen on the `pointer: ...` assignment).
Since this whole thing is already quite tricky/subtle, let's not risk unnecessary flexibility (which [appears to be unsound](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9ad1a40191c886a95e7d1e22ea3a0788)):

```suggestion
    Pin::<&mut _> { pointer: &mut { $value } }
```

Reply:
I've readded this and added a comment explaining why it should not be removed.

Reply:
I wonder if I should also add a test for this case in this PR, so that others cannot stumble over the same thing.